### PR TITLE
implement cdf for gamma distribution

### LINKED ISSUE #41637: Gamma distribution does not have a CDF implementation, which is very useful in Pyro
## 🚀 Feature
Implement a CDF method for the Gamma distribution (torch.distributions.gamma.Gamma)

## Motivation

I'm trying to write Pyro code which uses censored data on a Gamma distribution, but I need the CDF of the Gamma distribution for that. The CDF has closed form solution so it should be easy to implement. Currently I need to cheat by using Scipy's gamma distribution and its CDF so that incurs run time 

## Pitch

I envision a new method added to `torch.distributions.gamma.Gamma` with the signature

```
def cdf(self, value):
# closed form solution code here
```

However, this solution would require the gamma function which I'm not sure is implemented (is torch._standard_gamma the gamma function?). If it isn't implemented I'd have to use `torch.exp(torch.lgamma(...))`

## Alternatives

Currently I'm using Scipy's gamma distribution CDF. It's not great since the 
(a) the signature is different, and
(b) detaching, converting to numpy, and then back to pytorch tensors takes up run time

cc @vincentqb @fritzo @neerajprad @alicanb @vishwakftw

### PR DESCRIPTION
**Pull request by jan-glx**
Fixes #41637

Since I have no experience with pytorch and did not test this, this might be better seen as a "bump" to above issue.

### CODE REVIEW

**Comment by bdhirsh** [comment]
@mruberry Mostly adding you as a reviewer because it looks like nobody's touched this file in a while, and you shepherded another PR stemming from the originally linked issue: https://github.com/pytorch/pytorch/issues/41637. LMK if you think someone else would be better suited to take a look.

**Comment by mruberry** [comment]
Thanks @bdhirsh

**Comment by jan-glx** [comment]
Ok, The first commit's CI run showed that this is covered by 4 tests which all pass after actually implementing it correctly.
Thinking about how to best fix the style errors: Should I simply remove the comment?
Should I squash the commits?


Unrelated, for reference: implementation of Gamma's cdf in R:
https://github.com/wch/r-source/blob/cc60d0b82df14ac83dba10da2a3dafe76fdd0cbb/src/nmath/pgamma.c#L642-L718

**Comment by mruberry** [comment]
> Ok, The first commit's CI run showed that this is covered by 4 tests which all pass after actually implementing it correctly.
> Thinking about how to best fix the style errors: Should I simply remove the comment?
> Should I squash the commits?

You can just push an update to the branch. No need to squash.

> Unrelated, for reference: implementation of Gamma's cdf in R:
> https://github.com/wch/r-source/blob/cc60d0b82df14ac83dba10da2a3dafe76fdd0cbb/src/nmath/pgamma.c#L642-L718

**Comment by mruberry** [comment]
Thanks for submitting this PR, @jan-glx. You mention some tests covering the proper implementation of this cdf? Would you point me to them?

File: torch/distributions/gamma.py [hunk 1]
```diff
@@ -70,6 +70,11 @@ def log_prob(self, value):
                 (self.concentration - 1) * torch.log(value) -
                 self.rate * value - torch.lgamma(self.concentration))
 
+    def cdf(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        return torch.igamma(self.concentration, self.rate * value) # depsite its name, torch.igamma is the regularized gamma function
```
Comment (mruberry) [inline torch/distributions/gamma.py:76]: cc @kshitij12345 to also take a look. 

What name did you expect this function to have, @jan-glx? We're implementing the torch.special module now with updated names.

**Comment by jan-glx** [comment]
>
>
> You can just push an update to the branch. No need to squash.

OK. (Did that.)

> 
> 
> Thanks for submitting this PR, @jan-glx. You mention some tests covering the proper implementation of this cdf? Would you point me to them?

From[ the first commit's CI run](https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm4.0.1-py3.6-test2/6295/console):

> 12:55:21 ======================================================================
12:55:21 ERROR [0.060s]: test_cdf (__main__.TestAgainstScipy)
12:55:21 ----------------------------------------------------------------------
12:55:21 Traceback (most recent call last):
12:55:21   File "distributions/test_distributions.py", line 4378, in test_cdf
12:55:21     cdf = pytorch_dist.cdf(samples)
12:55:21   File "/opt/conda/lib/python3.6/site-packages/torch/distributions/gamma.py", line 76, in cdf
12:55:21     return torch.igamma(self.concentration, self.rate * value) / torch.gamma(self.concentration)
12:55:21 AttributeError: module 'torch' has no attribute 'gamma'
12:55:21 
12:55:21 ======================================================================
12:55:21 ERROR [0.008s]: test_cdf_icdf_inverse (__main__.TestDistributions)
12:55:21 ----------------------------------------------------------------------
12:55:21 Traceback (most recent call last):
12:55:21   File "distributions/test_distributions.py", line 2666, in test_cdf_icdf_inverse
12:55:21     cdf = dist.cdf(samples)
12:55:21   File "/opt/conda/lib/python3.6/site-packages/torch/distributions/gamma.py", line 76, in cdf
12:55:21     return torch.igamma(self.concentration, self.rate * value) / torch.gamma(self.concentration)
12:55:21 AttributeError: module 'torch' has no attribute 'gamma'
12:55:21 
12:55:21 ======================================================================
12:55:21 ERROR [0.006s]: test_cdf_log_prob (__main__.TestDistributions)
12:55:21 ----------------------------------------------------------------------
12:55:21 Traceback (most recent call last):
12:55:21   File "distributions/test_distributions.py", line 2687, in test_cdf_log_prob
12:55:21     cdfs = dist.cdf(samples)
12:55:21   File "/opt/conda/lib/python3.6/site-packages/torch/distributions/gamma.py", line 76, in cdf
12:55:21     return torch.igamma(self.concentration, self.rate * value) / torch.gamma(self.concentration)
12:55:21 AttributeError: module 'torch' has no attribute 'gamma'
12:55:21 
12:55:21 ======================================================================
12:55:21 ERROR [0.032s]: test_cdf (__main__.TestJit)
12:55:21 ----------------------------------------------------------------------
12:55:21 Traceback (most recent call last):
12:55:21   File "distributions/test_distributions.py", line 4810, in test_cdf
12:55:21     traced_f = torch.jit.trace(f, (sample,) + values)
12:55:21   File "/opt/conda/lib/python3.6/site-packages/torch/jit/_trace.py", line 785, in trace
12:55:21     get_callable_argument_names(func)
12:55:21   File "distributions/test_distributions.py", line 4806, in f
12:55:21     cdf = dist.cdf(sample)
12:55:21   File "/opt/conda/lib/python3.6/site-packages/torch/distributions/gamma.py", line 76, in cdf
12:55:21     return torch.igamma(self.concentration, self.rate * value) / torch.gamma(self.concentration)
12:55:21 AttributeError: module 'torch' has no attribute 'gamma'
12:55:21 
12:55:21 ----------------------------------------------------------------------

So, for example, [this one](https://github.com/pytorch/pytorch/blob/55dfb4a575746dc853c48591363ac4c4b8e7a065/test/distributions/test_distributions.py#L4374) compares the cdf with scipy's and was previously skipped.



--------

There is just one test failing now:
>
>
>14:17:39 ======================================================================
14:17:39 ERROR [0.014s]: test_cdf_log_prob (__main__.TestDistributions)
14:17:39 ----------------------------------------------------------------------
14:17:39 Traceback (most recent call last):
14:17:39   File "distributions/test_distributions.py", line 2691, in test_cdf_log_prob
14:17:39     cdfs_derivative = grad(cdfs.sum(), [samples])[0]  # this should not be wrapped in torch.abs()
14:17:39   File "/opt/conda/lib/python3.6/site-packages/torch/autograd/__init__.py", line 226, in grad
14:17:39     inputs, allow_unused, accumulate_grad=False)
14:17:39 RuntimeError: the derivative for 'igamma: input' is not implemented.
14:17:39 
14:17:39 ----------------------------------------------------------------------
14:17:39 Ran 207 tests in 40.960s
14:17:39 
14:17:39 FAILED (errors=1)

which might be some work to fix:
https://www.wolframalpha.com/input/?i=d+GammaRegularized%5Ba%2C+0%2C+z%5D+%2F+dz+%2C+d%2Fda+GammaRegularized%5Ba%2C+0%2C+z%5D

File: torch/distributions/gamma.py (hunk 1 above)
Comment (jan-glx) [reply torch/distributions/gamma.py:76]: Apparently there is quite some inconsistency across fields/programming languages:

| language | upper (Γ) - and |lower incomplete gamma function (γ) | regularized lower gamma function (P=γ/Γ) | 
| ---- | -----| ------| ------|
|[Wikipedia](https://en.wikipedia.org/wiki/Incomplete_gamma_function#Regularized_Gamma_functions_and_Poisson_random_variables)| upper incomplete gamma function | lower incomplete gamma function | regularized gamma function |
|[dlmf](https://dlmf.nist.gov/8.2#E4) | incomplete gamma function γ|  incomplete gamma function Γ | Normalized incomplete gamma function| 
| [julia](https://juliamath.github.io/SpecialFunctions.jl/dev/functions_list/#SpecialFunctions.gamma_inc) | gamma(a,x)|       | gamma_inc(a,x)[1] | 
| [scipy](https://docs.scipy.org/doc/scipy/reference/special.html#gamma-and-related-functions) |  |  | gammainc(a, x) |
| [GNU SF](https://www.gnu.org/software/gsl/doc/html/specfunc.html?highlight=gamma#incomplete-gamma-functions)| gsl_sf_gamma_inc | | gsl_sf_gamma_inc_P |  
| [Wolfram](https://reference.wolfram.com/language/tutorial/MathematicalFunctions.html#21909) | Gamma[a, z]|  Gamma[a,0,z1] | GammaRegularized |
| R |  |  | pgamma(z, a) |
| [TF](https://www.tensorflow.org/api_docs/python/tf/math/igamma) | | | tf.math.igamma |
| [MATLAB](https://de.mathworks.com/help/symbolic/sym.igamma.html#bt6_p8p-1) |  gammainc(x, a, 'lower')| igamma(a, x) or gammainc(x, a, 'upper') | gammainc (x,a,scale='scaledlower' )|

Not sure what is best, but igamma could also mean inverse gamma, so prob. not ideal.
While not indicating the normalization, id probably just stick with scipy....

Comment (kshitij12345) [reply torch/distributions/gamma.py:76]: Thanks for the detailed explanation.

> Not sure what is best, but igamma could also mean inverse gamma, so prob. not ideal.

I understand your concern now. As @mruberry mentioned, the special module will update the name of `igamma` to being consistent with SciPy's (i.e. `gammainc`)

Reference to the tracker issue: https://github.com/pytorch/pytorch/issues/50345

Comment (mruberry) [reply torch/distributions/gamma.py:76]: Thank you for sharing your perspective, @jan-glx.

Comment (jan-glx) [reply torch/distributions/gamma.py]: (removed the comment now, otherwise it will probably be forgotten after the rename of the specials)

**Comment by mruberry** [comment]
Looks like this test is failing on the ROCm build, and for some reason the rest of the PR CI didn't run. Maybe pushing an update will trigger it properly?

```
14:28:38 ======================================================================
14:28:38 ERROR [0.013s]: test_cdf_log_prob (__main__.TestDistributions)
14:28:38 ----------------------------------------------------------------------
14:28:38 Traceback (most recent call last):
14:28:38   File "distributions/test_distributions.py", line 2691, in test_cdf_log_prob
14:28:38     cdfs_derivative = grad(cdfs.sum(), [samples])[0]  # this should not be wrapped in torch.abs()
14:28:38   File "/opt/conda/lib/python3.6/site-packages/torch/autograd/__init__.py", line 226, in grad
14:28:38     inputs, allow_unused, accumulate_grad=False)
14:28:38 RuntimeError: the derivative for 'igamma: input' is not implemented.
14:28:38 
14:28:38 ----------------------------------------------------------------------
```

**Comment by nicola-decao** [comment]
Hi @jan-glx is it possible to obtain the CDF of the Beta distribution in a similar manner? Pytorch does not implement `betainc`  or `ibeta`

**Comment by kshitij12345** [comment]
> Pytorch does not implement betainc or ibeta

@nicola-decao , we are planning to add `betainc` soon as part of special module.

Tracking Issue: https://github.com/pytorch/pytorch/issues/50345

**Comment by nicola-decao** [comment]
@kshitij12345 that's cool!  But how soon? That issue is open since January and `betainc` is not ported yet. Is there any ETA? Just curious.

**Comment by kshitij12345** [comment]
@nicola-decao don't have an exact ETA but since there is interest will prioritise it :)

**Comment by jan-glx** [comment]
I just brought the branch up to date (by merging master into it, can do a re-base if necessary).
There is still just the `test_cdf_log_prob` test failing (for the not implemented derivative of igamma). To avoid this one could skip the test by moving the [gradient call](https://github.com/pytorch/pytorch/blob/55dfb4a575746dc853c48591363ac4c4b8e7a065/test/distributions/test_distributions.py#L2691) into the try block above.

**Pull request closed without merging**